File: algo/src/main/java/com/hartwig/actin/algo/evaluation/othercondition/HasSevereConcomitantIllness.java
package com.hartwig.actin.algo.evaluation.othercondition;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;

import org.jetbrains.annotations.NotNull;

public class HasSevereConcomitantIllness implements EvaluationFunction {

    HasSevereConcomitantIllness() {
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Integer whoStatus = record.clinical().clinicalStatus().who();

        if (whoStatus == null) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.UNDETERMINED)
                    .addUndeterminedSpecificMessages("Cannot determine whether patient has concomitant illnesses - who status unknown")
                    .addUndeterminedGeneralMessages("Unclear concomitant illnesses")
                    .build();
        }

        if (whoStatus >= 3) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.WARN)
                    .addWarnSpecificMessages("Patient has potential concomitant illnesses based on WHO status of " + whoStatus)
                    .addWarnGeneralMessages("Potential concomitant illnesses")
                    .build();
        }

        return EvaluationFactory.unrecoverable()
                .result(EvaluationResult.NOT_EVALUATED)
                .addPassSpecificMessages("Any severe concomitant illnesses are assumed not to be present")
                .addPassGeneralMessages("Assumed no severe concomitant illnesses")
                .build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/othercondition/HasSevereConcomitantIllness.kt
package com.hartwig.actin.algo.evaluation.othercondition

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation

class HasSevereConcomitantIllness: EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val whoStatus = record.clinicalStatus.who

        if (whoStatus == 3 || whoStatus == 4) {
            return EvaluationFactory.warn("Potential severe concomitant illnesses (WHO $whoStatus)")
        }
        return if (whoStatus == 5) {
            EvaluationFactory.pass("Has severe concomitant illness (WHO 5)")
        } else
            EvaluationFactory.notEvaluated("Assumed that severe concomitant illnesses are not to be present")
    }
}